DOCS-1467, DOCS-1480: Add Build and Platform page content (#2362)
viamrobotics/docs: docs/build/_index.md

---
title: "Build"
linkTitle: "Build"
description: "Configure and program your smart machine with Viam."
image: "/services/icons/sdk.svg"
imageAlt: "Tools and building blocks."
weight: 410
type: "docs"
no_list: true
menuindent: true
---

> Viam uses a composable system of building blocks called {{< glossary_tooltip term_id="resource" text="resources" >}} (components and services) so you can add the hardware drivers and the high-level software functionality you need to your smart machine:

Reviewer: Maybe this? I've removed the `(components and services)` because that's covered in the glossary tooltip. I think I don't like the "so you can add" - I think it's built that way to be composable in many different ways but not so the reader can add things. So that's what I'm trying to get towards:
```suggestion
Viam uses a composable system of building blocks called {{< glossary_tooltip term_id="resource" text="resources" >}} which you can combine according to your individual needs.
Components make your machine components usable with a consistent API and services provide high-level software funcitonality.
```

Author: ```suggestion
Viam uses a composable system of building blocks called {{< glossary_tooltip term_id="resource" text="resources" >}} which you can combine according to your specific needs.
Components drive your hardware and services provide high-level software functionality.
```
How about this? Don't want to say "components" twice in a row.

Reviewer: SGTM

<div class="cards max-page">
  <div class="row">
    <div class="col sectionlist">
        <div>
        <h2>Components:</h2>
        {{<sectionlist section="/components/">}}
        </div>
    </div>
    <div class="col sectionlist">

> <div><h2>Services:</h2><ul class="sectionlist"><li><a href="/data/" title="Data Management Service"><div><picture><img src="/services/icons/data-capture.svg" width="/services/icons/data-capture" height="/services/icons/data-capture" alt="Data Management" loading="lazy"></picture><p>Data Management</p></div></a></li></ul><ul class="sectionlist"><li><a href="/mobility/motion/" title="Motion Service"><div><picture><img src="/services/icons/motion.svg" width="/services/icons/motion" height="/services/icons/motion" alt="Motion" loading="lazy"></picture><p>Motion</p></div></a></li></ul><ul class="sectionlist"><li><a href="/mobility/frame-system/" title="Robot Frame System"><div><picture><img src="/services/icons/frame-system.svg" width="/services/icons/frame-system" height="/services/icons/frame-system" alt="Frame System" loading="lazy"></picture><p>Frame System</p></div></a></li></ul><ul class="sectionlist"><li><a href="/mobility/base-rc/" title="Base Remote Control Service"><div><picture><img src="/services/icons/base-rc.svg" width="/services/icons/base-rc" height="/services/icons/base-rc" alt="Base Remote Control" loading="lazy"></picture><p>Base Remote Control</p></div></a></li></ul><ul class="sectionlist"><li><a href="/ml/" title="ML Model Service"><div><picture><img src="/services/icons/ml.svg" width="/services/icons/ml" height="/services/icons/ml" alt="ML Model" loading="lazy"></picture><p>ML Model</p></div></a></li></ul><ul class="sectionlist"><li><a href="/mobility/navigation/" title="Navigation Service"><div><picture><img src="/services/icons/navigation.svg" width="/services/icons/navigation" height="/services/icons/navigation" alt="Navigation" loading="lazy"></picture><p>Navigation</p></div></a></li></ul><ul class="sectionlist"><li><a href="/mobility/slam/" title="SLAM Service"><div><picture><img src="/services/icons/slam.svg" width="/services/icons/slam" height="/services/icons/slam" alt="SLAM" loading="lazy"></picture><p>SLAM</p></div></a></li></ul><ul class="sectionlist"><li><a href="/ml/vision/" title="Vision Service"><div><picture><img src="/services/icons/vision.svg" width="/services/icons/vision" height="/services/icons/vision" alt="Vision" loading="lazy"></picture><p>Vision</p></div></a></li></ul></div>

Reviewer: Did you manually add a slash in front of each image src? I think that's the issue - I think it then cannot find the image
```suggestion
<div><div>Services</div><ul class="sectionlist"><li><a href="data/" title="Data Management Service"><div><picture><img src="services/icons/data-capture.svg" width="/services/icons/data-capture" height="/services/icons/data-capture" alt="Data Management" loading="lazy"></picture><p>Data Management</p></div></a></li></ul><ul class="sectionlist"><li><a href="mobility/motion/" title="Motion Service"><div><picture><img src="services/icons/motion.svg" width="/services/icons/motion" height="/services/icons/motion" alt="Motion" loading="lazy"></picture><p>Motion</p></div></a></li></ul><ul class="sectionlist"><li><a href="mobility/frame-system/" title="The Robot Frame System"><div><picture><img src="services/icons/frame-system.svg" width="/services/icons/frame-system" height="/services/icons/frame-system" alt="Frame System" loading="lazy"></picture><p>Frame System</p></div></a></li></ul><ul class="sectionlist"><li><a href="mobility/base-rc/" title="Base Remote Control Service"><div><picture><img src="services/icons/base-rc.svg" width="/services/icons/base-rc" height="/services/icons/base-rc" alt="Base Remote Control" loading="lazy"></picture><p>Base Remote Control</p></div></a></li></ul><ul class="sectionlist"><li><a href="ml/" title="ML Model Service"><div><picture><img src="services/icons/ml.svg" width="/services/icons/ml" height="/services/icons/ml" alt="ML Model" loading="lazy"></picture><p>ML Model</p></div></a></li></ul><ul class="sectionlist"><li><a href="mobility/navigation/" title="The Navigation Service"><div><picture><img src="services/icons/navigation.svg" width="/services/icons/navigation" height="/services/icons/navigation" alt="Navigation" loading="lazy"></picture><p>Navigation</p></div></a></li></ul><ul class="sectionlist"><li><a href="mobility/slam/" title="SLAM Service"><div><picture><img src="services/icons/slam.svg" width="/services/icons/slam" height="/services/icons/slam" alt="SLAM" loading="lazy"></picture><p>SLAM</p></div></a></li></ul><ul class="sectionlist"><li><a href="ml/vision/" title="Vision Service"><div><picture><img src="services/icons/vision.svg" width="/services/icons/vision" height="/services/icons/vision" alt="Vision" loading="lazy"></picture><p>Vision</p></div></a></li></ul></div>
```

Author: I did because locally, it didn't work without the slash.

Reviewer: Yeah my solution is wrong, this needs to use relative links which are different if they come from the build page:
```suggestion
<div><div>Services</div><ul class="sectionlist"><li><a href="data/" title="Data Management Service"><div><picture><img src="../services/icons/data-capture.svg"   alt="Data Management" loading="lazy"></picture><p>Data Management</p></div></a></li></ul><ul class="sectionlist"><li><a href="mobility/motion/" title="Motion Service"><div><picture><img src="../services/icons/motion.svg" alt="Motion" loading="lazy"></picture><p>Motion</p></div></a></li></ul><ul class="sectionlist"><li><a href="mobility/frame-system/" title="The Robot Frame System"><div><picture><img src="../services/icons/frame-system.svg" alt="Frame System" loading="lazy"></picture><p>Frame System</p></div></a></li></ul><ul class="sectionlist"><li><a href="mobility/base-rc/" title="Base Remote Control Service"><div><picture><img src="../services/icons/base-rc.svg" alt="Base Remote Control" loading="lazy"></picture><p>Base Remote Control</p></div></a></li></ul><ul class="sectionlist"><li><a href="ml/" title="ML Model Service"><div><picture><img src="../services/icons/ml.svg" alt="ML Model" loading="lazy"></picture><p>ML Model</p></div></a></li></ul><ul class="sectionlist"><li><a href="mobility/navigation/" title="The Navigation Service"><div><picture><img src="../services/icons/navigation.svg" alt="Navigation" loading="lazy"></picture><p>Navigation</p></div></a></li></ul><ul class="sectionlist"><li><a href="mobility/slam/" title="SLAM Service"><div><picture><img src="../services/icons/slam.svg" alt="SLAM" loading="lazy"></picture><p>SLAM</p></div></a></li></ul><ul class="sectionlist"><li><a href="ml/vision/" title="Vision Service"><div><picture><img src="../services/icons/vision.svg" alt="Vision" loading="lazy"></picture><p>Vision</p></div></a></li></ul></div>
```

Author: Ah and now the hrefs need to be absolute links but the relative icons work! cool

    </div>
  </div>
</div>

> For additional flexibility see the [modular resource registry](/registry/).

Author: ```suggestion
If a component or service you want to use for your project is not natively supported, see the [modular resource registry](/registry/).
```

## SDKs

Each category of resource has a standardized API that you can access with an SDK (software development kit) in your preferred programming language.
For example, you can send the same commands to any kind of motor, using any of the following programming languages:

{{<sectionlist section="/sdks">}}

## Start building

To get started, first [install `viam-server`](/get-started/installation/) on your smart machine's computer.
Viam supports a wide variety of popular systems, including:

> <div id="board-carousel" class="carousel max-page">

Reviewer: I like the Platform and the Build pages overall - I am concerned with the placement of the carousel and am wondering if we should move this up on the page. Start with that, then the components and services, and then programming. I think you can probably do that and then have the two cards still under a Next Steps header at the bottom - what do you think?

Author: Here, I wanted to emphasize the "building blocks" concept before getting into specific hardware, and before getting into the specific "board" component.

Author: Okay, and also leave it here? Or move it out of here?

Reviewer: Move out of here there - you did say it didn't fit here right?

Author: I mean that I think it fits at the bottom of build but not at the top of build.

        <p>PCA9686 I<sup>2</sup>C Interface</p>
      </a>
    </li>
  </ul>
  <ol style="visibility: hidden" aria-hidden="true">
    <li><a href="#c1_slide0">Raspberry Pi 5</a></li>
    <li><a href="#c1_slide1">Raspberry Pi 4</a></li>
    <li><a href="#c1_slide2">NVIDIA Jetson Orin Nano</a></li>
    <li><a href="#c1_slide3">Intel UP4000</a></li>
    <li><a href="#c1_slide4">Mediatek Genio 500 Pumpkin</a></li>
    <li><a href="#c1_slide5">Espressif ESP32</a></li>
    <li><a href="#c1_slide6">Texas Instruments TDA4VM</a></li>
    <li><a href="#c1_slide7">NVIDIA Jetson Nano</a></li>
    <li><a href="#c1_slide8">NVIDIA Jetson AGX Orin</a></li>
    <li><a href="#c1_slide9">NVIDIA Jetson Xavier NX</a></li>
    <li><a href="#c1_slide10">Raspberry Pi 3</a></li>
    <li><a href="#c1_slide11">Raspberry Pi Zero 2 W</a></li>
    <li><a href="#c1_slide12">BeagleBone AI-64</a></li>
    <li><a href="#c1_slide13">Numato GPIO Modules</a></li>
    <li><a href="#c1_slide14">PCA9686 I<sup>2</sup>C Interface</a></li>
  </ol>
  <div class="prev" style="display: block">‹</div>
  <div class="next" style="display: block">›</div>
</div>
<br>

> Then pull together whichever Viam resources you like by configuring your smart machine.
> Finally, program your smart machine with an SDK in your preferred coding language.

Reviewer: Let's combine this and then just have the cards do the rest
```suggestion
Then pull together the Viam resources you need by configuring your smart machine before programming your machine:
```

Author: That feels like a lot in a single sentence. Also I was hoping to provide a little more of a step-by-step call to action situation.

I could move the board carousel out of here onto the platform page, and turn the "Start building" section into an ordered list of 3 steps. Each step would have a link, and the cards would be removed entirely.

Reviewer: my issue with it is that there's two sentences that feel like they should have links but don't because the cards are immediately after. You can keep the two sentences if you prefer but you do need to get rid of the `whichever` and `you like`, that is a bit too colloquial

Author: ```suggestion
Then pull together the Viam resources you want to use by configuring your smart machine.
Finally, program your smart machine with an SDK in your preferred coding language.
```
